### PR TITLE
152 - Swaplist component search feature

### PR DESCRIPTION
**Pull request by ericangeles**
**Explain the _details_ for making this change. What existing problem does the pull request solve?**

Additional feature for swaplist. 

**Related github/jira issue (required)**:

Closes https://github.com/infor-design/enterprise-ng/issues/152

**Steps necessary to review your pull request (required)**:

- Run http://localhost:4200/swaplist-search
- Searchfield should be visible and functioning when searching items

### CODE REVIEW

**Comment by bthharper** [comment]
Easy one:

> ERROR: /home/travis/build/infor-design/enterprise-ng/src/app/swaplist/swaplist-search.demo.ts[85, 1]: Consecutive blank lines are forbidden

File: projects/ids-enterprise-ng/src/lib/swaplist/soho-swaplist.component.ts [hunk 1]
```diff
@@ -148,6 +179,18 @@ export class SohoSwapListComponent implements AfterViewInit, OnDestroy {
   @Input()
   public fullAccessCardTitle = Soho.Locale.translate('AdditionalItems');
 
+  /** Default search title for available items card. */
+  @Input()
```
Comment (bthharper) [inline projects/ids-enterprise-ng/src/lib/swaplist/soho-swaplist.component.ts]: Potentially unnecessary locale look ups?  Maybe add a getter which return the default if non provided?

File: projects/ids-enterprise-ng/src/lib/swaplist/soho-swaplist.component.ts [hunk 2]
```diff
@@ -161,17 +204,36 @@ export class SohoSwapListComponent implements AfterViewInit, OnDestroy {
...
-    *
-    * @param value item data.
-    */
+   * Set searchable option for SohoSwapList.
+   *
+   * @param value option.
+   */
+  @Input()
+  public set searchable(value: boolean) {
+    this._options.searchable = value;
```
Comment (bthharper) [inline projects/ids-enterprise-ng/src/lib/swaplist/soho-swaplist.component.ts]: Does this needs to call `updated` on the widget if this option is changed?

File: projects/ids-enterprise-ng/src/lib/swaplist/soho-swaplist.component.ts [hunk 3]
```diff
@@ -271,24 +333,25 @@ export class SohoSwapListComponent implements AfterViewInit, OnDestroy {
...
+        this.updateDataset(d);
+      });
     }
   }
 
   /**
-  * Destroys any resources created by the control.
-  */
+   * Destroys any resources created by the control.
+   */
```
Comment (bthharper) [inline projects/ids-enterprise-ng/src/lib/swaplist/soho-swaplist.component.ts]: Add `this.jqueryElement.off()`

File: projects/ids-enterprise-ng/src/lib/swaplist/soho-swaplist.component.ts [hunk 4]
```diff
@@ -271,24 +333,25 @@ export class SohoSwapListComponent implements AfterViewInit, OnDestroy {
 
   ngAfterViewInit() {
     this.jQueryElement = jQuery(this.element.nativeElement);
+    this.jQueryElement.find('.searchfield').searchfield({ clearable: true });
```
Comment (bthharper) [inline projects/ids-enterprise-ng/src/lib/swaplist/soho-swaplist.component.ts]: Should this not be an option?

**Comment by bthharper** [comment]
Could we add a unit test in to?

File: projects/ids-enterprise-ng/src/lib/swaplist/soho-swaplist.component.ts (hunk 1 above)
Comment (pwpatton) [reply projects/ids-enterprise-ng/src/lib/swaplist/soho-swaplist.component.ts]: We made a translate pipe in landmark that will take 'AdditionalTimes | translate' in the template and run the pipe class against that. Apparently it's supposed to be more efficient too. Perhaps this should be in ep-ng?

```
import { Pipe, PipeTransform } from '@angular/core';

@Pipe({
  name: 'translate'
})
export class TranslatePipe implements PipeTransform {
  transform(value: string): string {
    return Soho.Locale.translate(value);
  }
}
```

**Comment by pwpatton** [comment]
`@pwpatton would this be a good time to add the ngZone code too ?`

Yea, we should definitely do it now.

File: projects/ids-enterprise-ng/src/lib/swaplist/soho-swaplist.component.ts (hunk 1 above)
Comment (bthharper) [reply projects/ids-enterprise-ng/src/lib/swaplist/soho-swaplist.component.ts]: This is baked into `ids-enterprise-ng` as sohoTranslate.  

```html
 <div style="text-align: center;">
  <button soho-button="primary">{{'Yes'|sohoTranslate}}</button>
  <button soho-button="secondary">{{'No'|sohoTranslate}}</button>
</div>
```

There are also date and number formatters.

```html
 <div class="field">
   <input soho-textarea value="{{model.number|sohoFormatNumber}}"/>
   <input soho-textarea value="{{model.number|sohoFormatNumber:{style: 'currency'} }}"/>
  <input soho-textarea  value="{{model.number|sohoFormatNumber:{style: 'percent'} }}"/>
</div>
```

**Comment by tmcconechy** [comment]
Do the ngZone changes supposed to be added to this PR? Or will it be done on the other one?

File: projects/ids-enterprise-ng/src/lib/swaplist/soho-swaplist.component.ts [hunk 5]
```diff
@@ -101,21 +132,21 @@ export class SohoSwapListComponent implements AfterViewInit, OnDestroy {
   private _showFullAccessCard = false;
```
Comment (tmcconechy) [inline projects/ids-enterprise-ng/src/lib/swaplist/soho-swaplist.component.ts]: From a stylistic side are we consistently naming the variables that are internal with an underscore? I feel like its mixed and probably should just drop the underscore @bthharper  @pwpatton ?

Comment (bthharper) [reply projects/ids-enterprise-ng/src/lib/swaplist/soho-swaplist.component.ts]: Yes - no underscores (or Hungarian notation).

**Comment by bthharper** [comment]
> Do the ngZone changes supposed to be added to this PR? Or will it be done on the other one?

I don't think its included in @pwpatton work, would be great to do it in this one.  Appreciate that it is moving the goal posts somewhat.

File: projects/ids-enterprise-ng/src/lib/swaplist/soho-swaplist.component.ts (hunk 1 above)
Comment (ericangeles) [reply projects/ids-enterprise-ng/src/lib/swaplist/soho-swaplist.component.ts]: I tried to remove all the `Soho.Locale.translate` and added the `{{'AdditionalItems'|sohoTranslate}}` to the label and input field but it seems the `sohoTranslate` couldn't find. I got an error after executing the `npm run build:lib`

Comment (bthharper) [reply projects/ids-enterprise-ng/src/lib/swaplist/soho-swaplist.component.ts]: You may need to import the Locale pipe into the module.

Comment (ericangeles) [reply projects/ids-enterprise-ng/src/lib/swaplist/soho-swaplist.component.ts]: Still getting the error.

**Pull request merged**